File: lib/commands/pr.js
const fs = require("fs");
const path = require("path");
const { execSync } = require("child_process");
const { gitInBare } = require("../git");
const { loadConfig } = require("../config");
const { setupWorktree } = require("./create");

function pr(number) {
  // Get PR branch name from GitHub CLI
  let branch;
  try {
    branch = execSync(
      `gh pr view ${number} --json headRefName -q .headRefName`,
      { encoding: "utf-8", stdio: ["pipe", "pipe", "pipe"] }
    ).trim();
  } catch {
    console.error(`Failed to get PR #${number}. Is \`gh\` installed and authenticated?`);
    process.exit(1);
  }

  const { root, config } = loadConfig();
  const { remote, defaultBase } = config;
  const base = defaultBase;
  const prsDir = path.join(root, "prs");
  const worktreePath = path.join(prsDir, String(number));

  if (fs.existsSync(worktreePath)) {
    console.error(`Directory "prs/${number}" already exists.`);
    process.exit(1);
  }

  // Fetch latest from remote
  console.log(`Fetching from ${remote}...`);
  gitInBare(`fetch ${remote}`, root);

  // Check if branch already exists locally
  let branchExists = false;
  try {
    gitInBare(`rev-parse --verify refs/heads/${branch}`, root);
    branchExists = true;
  } catch {}

  // Ensure prs/ directory exists
  fs.mkdirSync(prsDir, { recursive: true });

  // Create worktree from the PR's remote branch
  console.log(`Creating worktree for PR #${number} (${branch}) from ${remote}/${branch}...`);
  if (branchExists) {
    gitInBare(`worktree add ${worktreePath} ${branch}`, root);
  } else {
    gitInBare(
      `worktree add ${worktreePath} -b ${branch} --no-track ${remote}/${branch}`,
      root
    );
  }

  setupWorktree(root, config, worktreePath, base);

  console.log(`\nWorktree ready at ./prs/${number}`);
  console.log(`  PR: #${number}`);
  console.log(`  Branch: ${branch}`);

  console.log(`\n  cd prs/${number}`);
}

module.exports = pr;

File: internal/cmd/pr.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
)

func PrCmd(d *Deps) *cobra.Command {
	return &cobra.Command{
		Use:   "pr <number>",
		Short: "Create a worktree for a pull request",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			number := args[0]

			// Get PR branch name from GitHub CLI
			out, err := exec.Command("gh", "pr", "view", number, "--json", "headRefName", "-q", ".headRefName").Output()
			if err != nil {
				return fmt.Errorf("failed to get PR #%s. Is `gh` installed and authenticated?", number)
			}
			branch := strings.TrimSpace(string(out))

			root, cfg, err := d.Config.LoadConfig()
			if err != nil {
				return err
			}

			base := cfg.DefaultBase
			prsDir := filepath.Join(root, "prs")
			worktreePath := filepath.Join(prsDir, number)

			if fileOrDirExists(worktreePath) {
				return fmt.Errorf("directory \"prs/%s\" already exists", number)
			}

			// Fetch latest from remote
			fmt.Fprintf(d.Stdout, "Fetching from %s...\n", cfg.Remote)
			if _, err := d.Git.GitInBare("fetch "+cfg.Remote, root); err != nil {
				return fmt.Errorf("fetch failed: %w", err)
			}

			// Delete stale local branch if it exists
			if _, err := d.Git.GitInBare("rev-parse --verify refs/heads/"+branch, root); err == nil {
				fmt.Fprintf(d.Stdout, "Deleting stale local branch %q...\n", branch)
				if _, err := d.Git.GitInBare("branch -D "+branch, root); err != nil {
					return fmt.Errorf("failed to delete stale branch: %w", err)
				}
			}

			// Ensure prs/ directory exists
			if err := os.MkdirAll(prsDir, 0755); err != nil {
				return err
			}

			// Create worktree from the PR's remote branch
			fmt.Fprintf(d.Stdout, "Creating worktree for PR #%s (%s) from %s/%s...\n", number, branch, cfg.Remote, branch)
			gitArgs := fmt.Sprintf("worktree add %s -b %s --no-track %s/%s", worktreePath, branch, cfg.Remote, branch)
			if _, err := d.Git.GitInBare(gitArgs, root); err != nil {
				return fmt.Errorf("worktree add failed: %w", err)
			}

			if err := SetupWorktree(d, root, cfg, worktreePath, base); err != nil {
				return err
			}

			fmt.Fprintf(d.Stdout, "\nWorktree ready at ./prs/%s\n", number)
			fmt.Fprintf(d.Stdout, "  PR: #%s\n", number)
			fmt.Fprintf(d.Stdout, "  Branch: %s\n", branch)
			fmt.Fprintf(d.Stdout, "\n  cd prs/%s\n", number)
			return nil
		},
	}
}
